packages/dds/merge-tree/src/stamps.ts

```diff
@@ -0,0 +1,157 @@
+ *
+ * @remarks The terminology here comes from the fact that the removal should affect only the *set* of nodes that were
+ * specified at the time the local client issued the remove, and not any nodes that were inserted concurrently.
+ *
+ * Not using "remove" and "obliterate" here allows us to unambiguously use the term "remove" elsewhere in code to mean
+ * "removed from the tree, either by MergeTree.obliterateRange or MergeTree.removeRange". This is convenient as the vast majority
+ * of merge-tree code only cares about segment visibility and not the specific operation that caused a segment to be removed.
+ */
+export interface SetRemoveOperationStamp extends OperationStamp {
+	readonly type: "setRemove";
+}
+
+/**
+ * {@link OperationStamp} for a 'set remove' operation. This aligns with the `mapRangeRemoved` API in MergeTree.
+ *
+ * @remarks The terminology here comes from the fact that the removal should affect the *slice* of nodes between the
+ * start and end point specified by the local client, which includes any nodes that were inserted concurrently.
+ *
+ * Not using "remove" and "obliterate" here allows us to unambiguously use the term "remove" elsewhere in code to mean
+ * "removed from the tree, either by MergeTree.obliterateRange or MergeTree.removeRange". This is convenient as the vast majority
+ * of merge-tree code only cares about segment visibility and not the specific operation that caused a segment to be removed.
+ */
+export interface SliceRemoveOperationStamp extends OperationStamp {
+	readonly type: "sliceRemove";
+}
+
+export type RemoveOperationStamp = SetRemoveOperationStamp | SliceRemoveOperationStamp;
+
+export function lessThan(a: OperationStamp, b: OperationStamp): boolean {
+	if (a.seq === UnassignedSequenceNumber) {
+		return b.seq === UnassignedSequenceNumber && a.localSeq! < b.localSeq!;
+	}
+
+	if (b.seq === UnassignedSequenceNumber) {
+		return true;
+	}
+
+	return a.seq < b.seq;
+}
+
+export function gte(a: OperationStamp, b: OperationStamp): boolean {
+	return !lessThan(a, b);
+}
+
+export function greaterThan(a: OperationStamp, b: OperationStamp): boolean {
+	if (a.seq === UnassignedSequenceNumber) {
+		return b.seq !== UnassignedSequenceNumber || a.localSeq! > b.localSeq!;
+	}
+
+	if (b.seq === UnassignedSequenceNumber) {
+		return false;
+	}
+
+	return a.seq > b.seq;
+}
+
+export function lte(a: OperationStamp, b: OperationStamp): boolean {
+	return !greaterThan(a, b);
+}
+
+export function equal(a: OperationStamp, b: OperationStamp): boolean {
+	return a.seq === b.seq && a.clientId === b.clientId && a.localSeq === b.localSeq;
+}
+
+export function isLocal(a: OperationStamp) {
+	return a.seq === UnassignedSequenceNumber;
+}
+
+export function isAcked(a: OperationStamp) {
+	return a.seq !== UnassignedSequenceNumber;
+}
+
+/**
+ * Inserts a stamp into a sorted list of stamps in the correct (sorted) position.
+ */
+export function insertIntoList(list: OperationStamp[], stamp: OperationStamp): void {
```

Comment:
would a doubly linked list be better? is indexing important elsewhere? array splice is O(N), so if we insert in a loop we can pretty easily end up at O(N*M)

Reply:
The asymptotics end up ok because we only ever use this to add removal information for a given segment, so the insertion point ends up being either at the end or 1 off from the end (it's why I chose that iteration order).

Rather than use a different data structure, I'd rather just remove this as a utility and move the logic to merge-tree where it's more obvious our use case is scoped but put it somewhere we can still unit test.